src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs

```diff
@@ -1,882 +0,0 @@
-            if (this.ReplacementText == string.Empty)
-            {
-                return;
-            }
-
-            var replacementText = this.ReplacementText;
-            var optionSet = _optionSet;
-            var cancellationToken = _conflictResolutionTaskCancellationSource.Token;
-
-            var asyncToken = _asyncListener.BeginAsyncOperation(nameof(UpdateConflictResolutionTask));
-
-            _conflictResolutionTask = ThreadingContext.JoinableTaskFactory.RunAsync(async () =>
-            {
-                // Join prior work before proceeding, since it performs a required state update.
-                // https://github.com/dotnet/roslyn/pull/34254#discussion_r267024593
-                //
-                // If cancellation of the conflict resolution task is requested before the rename locations task
-                // completes, we do not need to wait for rename before cancelling. The next conflict resolution task
-                // will wait on the latest rename location task if/when necessary.
-                var result = await _allRenameLocationsTask.JoinAsync(cancellationToken).ConfigureAwait(false);
-                await TaskScheduler.Default;
-
-                return await result.GetReplacementsAsync(replacementText, optionSet, cancellationToken).ConfigureAwait(false);
-            });
-
-            _conflictResolutionTask.Task.CompletesAsyncOperation(asyncToken);
-        }
-
-        private void QueueApplyReplacements()
-        {
-            // If the replacement text is empty, we do not update the results of the conflict
-            // resolution task. We instead wait for a non-empty identifier.
-            if (this.ReplacementText == string.Empty)
-            {
-                return;
-            }
-
-            var asyncToken = _asyncListener.BeginAsyncOperation(nameof(QueueApplyReplacements));
-            _conflictResolutionTask.Task
-                .SafeContinueWith(
-                    t => ComputeMergeResultAsync(t.Result, _conflictResolutionTaskCancellationSource.Token),
-                    _conflictResolutionTaskCancellationSource.Token,
-                    TaskContinuationOptions.OnlyOnRanToCompletion,
-                    TaskScheduler.Default)
-                .Unwrap()
-                .SafeContinueWithFromAsync(
-                    async t =>
-                    {
-                        await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, _conflictResolutionTaskCancellationSource.Token);
-
-                        ApplyReplacements(t.Result.replacementInfo, t.Result.mergeResult, _conflictResolutionTaskCancellationSource.Token);
-                    },
-                    _conflictResolutionTaskCancellationSource.Token,
-                    TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
-                    TaskScheduler.Default)
-                .CompletesAsyncOperation(asyncToken);
```

Comment:
[Tagging this here because GitHub won't let me tag the current version.]

So oddly the Cocoa implementation has a different version of this which is using JTF to schedule this work:

https://github.com/sharwell/roslyn/blob/7de8bb93b5976c98200f87c468f6f276505a7b34/src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.cs#L578-L590

The result of this PR is taking the Cocoa one which is is obviously a lot cleaner but has a bug: ComputeMergeResultAsync should be run on a background thread as otherwise we're doing computationally expensive work on the UI thread that doesn't need to be there. In the Cocoa implementation however all the calls to QueueApplyReplacements are called on the UI thread (and asserted as such), and the awaits are all staying on the UI thread.

I think the fix is just to add an explicit `await TaskScheduler.Default` before this line:

https://github.com/sharwell/roslyn/blob/7de8bb93b5976c98200f87c468f6f276505a7b34/src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.cs#L586

Otherwise I worry we're carrying a (potentially pretty worrisome) perf issue from the Cocoa implementation into the WPF implementation.

Reply:
➡️  now fixed.